great_expectations/datasource/fluent/config.py

```diff
@@ -111,13 +107,25 @@ def _load_datasource_subtype(cls, v: Dict[str, dict]):
                     f"'{ds_name}' has unsupported 'type' - {type_lookup_err}"
                 ) from type_lookup_err
 
-            config["name"] = ds_name
+            if "name" in config:
+                if config["name"] != ds_name:
```

Comment:
```suggestion
            ds_name = config.get("name",  MISSING) # MISSING = object()
            if ds_name is not MISSING and ds_name != ds_key:
```

In the suggestion above `MISSING` is a global constant sentinel value that can be used to determine whether or not the `"name"` key is actually present.

Also in this example, I've renamed the existing `ds_name` to `ds_key` and use `ds_name` to represent the value of the `Datasource.name` if it's present.